File: 0981-time-based-key-value-store/0981-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.d = collections.defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.d[key].append((value, timestamp))

    def get(self, key: str, timestamp: int) -> str:
        pairs = self.d[key]
        result = ''
        
        # binary search
        l = 0
        r = len(pairs) - 1
        
        while l <= r:
            m = (l + r) // 2
            
            value, current = pairs[m][0], pairs[m][1]
            
            if current == timestamp:
                return value

            if current > timestamp:
                r = m - 1
            else:
                l = m + 1
                result = value
        
        return result

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.java
class TimeMap {
    // approach: hashmap of key : list of [val, timestamp]
    
    Map<String, List<Pair<String, Integer>>> hashmap;
    
    public TimeMap() {
        hashmap = new HashMap<>();
    }
    
    public void set(String key, String value, int timestamp) {
        if (!hashmap.containsKey(key)) {
            hashmap.put(key, new ArrayList<>());
        }
        
        hashmap.get(key).add(new Pair(value, timestamp));
    }
    
    public String get(String key, int timestamp) {
        if (!hashmap.containsKey(key)) {
            return "";
        }
        
        // binary search on timestamps
        List<Pair<String, Integer>> messages = hashmap.get(key);
        
        int l = 0;
        int r = messages.size() - 1;
        
        String result = "";

        while (l <= r) {
            
            int m = (l + r) / 2;
            
            int current = messages.get(m).getValue();
            
            if (current == timestamp) {
                return messages.get(m).getKey();
            }
            
            if (current > timestamp) {
                r = m - 1;
            } else {
                result = messages.get(m).getKey();
                l = m + 1;
            }   
        }
        
        return result;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */